password/src/Data/Password/Validate.hs

```diff
@@ -22,53 +22,71 @@ Validate module provides set of functions which enables you to validate them.
 -}
 
 module Data.Password.Validate
-  ( -- * Data type
+  ( -- * Validating passwords
+    validatePassword,
+    isValidPassword,
+    ValidationResult(..),
+    -- ** Password Policy
+    validatePasswordPolicy,
     PasswordPolicy (..),
+    ValidPasswordPolicy,
+    unValidatePasswordPolicy,
+    defaultPasswordPolicy,
+    defaultPasswordPolicy_,
+    CharSetPredicate(..),
+    defaultCharSetPredicate,
     InvalidReason (..),
     InvalidPolicyReason(..),
     CharacterCategory(..),
-    -- * Predicate
-    CharSetPredicate(..),
-    -- * Default values
-    defaultPasswordPolicy,
-    defaultCharSetPredicate,
-    -- * Functions
-    isValidPassword,
-    validatePassword,
+    MinimumLength,
+    MaximumLength,
+    ProvidedLength,
+    MinimumAmount,
+    ProvidedAmount,
     -- * For internal use
-    isValidPasswordPolicy,
-    isSpecial,
     defaultCharSet,
-    categoryToPredicate
+    validateCharSetPredicate,
+    categoryToPredicate,
+    isSpecial
   ) where
 
 import Data.Char (chr, isAsciiLower, isAsciiUpper, isDigit, ord)
-import Data.Maybe (catMaybes)
-import Data.Password.Internal (Password (..))
+import Data.Function (on)
 #if! MIN_VERSION_base(4,13,0)
 import Data.Semigroup ((<>))
 #endif
 import Data.Text (Text)
 import qualified Data.Text as T
 
+import Data.Password.Internal (Password (..))
+
 -- $setup
 -- >>> :set -XOverloadedStrings
 --
 -- Import needed libraries.
 --
 -- >>> import Data.Password
 
--- | Set of policies used to validate 'Password'
+{-
+TODO: Add a QuasiQuoter to check password policies at compile time.
```

Comment:
Let's make sure we don't forget to create an issue for this, since it would be something relatively easy for someone not familiar with this library to tackle.

Reply:
Maybe, the way in which to set up the QuasiQuoter would require some thought, though.

Would it look like this?
`[validateQQ|min=10 max=64 lower=2 upper=2 special=1 digit=1|]`

Something else? Do any fields not set use `defaultPasswordPolicy`? Or do we have different QQs for that? And what if you want to provide a custom `CharSetPredicate`?

Lots of choices to be made.

Reply:
I think you are right that a lot of different choices could be made here.

I was more thinking of a template Haskell constructor than a quasiquoter.  Something similar to the constructors in `path` (for example, https://hackage.haskell.org/package/path-0.8.0/docs/Path-Posix.html#v:mkAbsDir).

So you'd be able to write something like this:

```haskell
myPassPolicy :: ValidPasswordPolicy
myPassPolicy =
  $(validatePassPolicyTH
      defaultPasswordPolicy { minimumLength = 10, maximumLength = 100 }
    )
```

I imagine this would be sufficient for many use-cases (well, it may be a little inconvenient to specify a `charSetPredicate`), although having a quasiquoter may be nice as well.

Reply:
Oh right, that's probably a lot easier and straightforward!
I think I can add that in and finish up this whole addition next weekend.